File: impl_python/type_check/decorator.py
# git SHA: 08f19f554735e8588bc1f4bd2e2b300b43e4a31a
"""Decorator type checking mixin (mirrors src/type_check.rs)."""
from __future__ import annotations
from typing import TYPE_CHECKING

from ..ast import Expr, ExprIdent
from .types import TyAny, TyTypeVal, TyTypeValOf, TyUnresolved, TyFunction, InferredType
from .errors import ErrInvalidDecorator

if TYPE_CHECKING:
    pass


class _TypeCheckerDecorator:
    """Mixin providing decorator type checking for TypeChecker."""

    def _check_decorator(self, dec: Expr, is_fn_target: bool, target_name: str) -> None:
        if not isinstance(dec, ExprIdent):
            return
        dec_name = dec.name
        expected_arg = "function" if is_fn_target else "type"
        expected_ret = "function" if is_fn_target else "type"

        sigs = self._fn_sigs.get(dec_name)
        cls_sigs = self._class_method_sigs.get(dec_name)

        def _is_ok_arg(ty: "InferredType") -> bool:
            return (isinstance(ty, TyFunction) if expected_arg == "function"
                    else isinstance(ty, (TyTypeVal, TyTypeValOf)))

        def _is_ok_ret(ty: "InferredType") -> bool:
            return (isinstance(ty, TyFunction) if expected_ret == "function"
                    else isinstance(ty, (TyTypeVal, TyTypeValOf)))

        if sigs:
            if len(sigs) != 1:
                return
            sig = sigs[0]
            if len(sig.params) != 1:
                self._report(ErrInvalidDecorator(
                    reason=f"'{dec_name}' must take exactly 1 argument, takes {len(sig.params)}"
                ))
                return
            pty = sig.params[0][1]
            if pty and not isinstance(pty, (TyUnresolved, TyAny)) and not _is_ok_arg(pty):
                self._report(ErrInvalidDecorator(
                    reason=f"'{dec_name}' parameter must be '{expected_arg}'"
                ))
            rty = sig.return_type
            if rty and not isinstance(rty, (TyUnresolved, TyAny)) and not _is_ok_ret(rty):
                self._report(ErrInvalidDecorator(
                    reason=f"'{dec_name}' must return '{expected_ret}'"
                ))

        elif cls_sigs:
            init_sigs = cls_sigs.get("__init__")
            call_sigs = cls_sigs.get("__call__")
            if init_sigs and len(init_sigs) == 1:
                params = init_sigs[0].params
                if len(params) >= 2:
                    pty = params[1][1]
                    if pty and not isinstance(pty, (TyUnresolved, TyAny)) and not _is_ok_arg(pty):
                        self._report(ErrInvalidDecorator(
                            reason=f"'{dec_name}.__init__' second param must be '{expected_arg}'"
                        ))
            if call_sigs and len(call_sigs) == 1:
                rty = call_sigs[0].return_type
                if rty and not isinstance(rty, (TyUnresolved, TyAny)) and not _is_ok_ret(rty):
                    self._report(ErrInvalidDecorator(
                        reason=f"'{dec_name}.__call__' must return '{expected_ret}'"
                    ))

File: src/type_check/decorator.rs
use crate::ast::Expr;

use super::errors::{StaticTypeError, TypeErrorKind};
use super::types::InferredType;
use super::TypeChecker;

impl TypeChecker {
    /// デコレータ式の型シグネチャを検査する。
    pub(super) fn check_decorator(
        &mut self,
        decorator: &Expr,
        target_is_fn: bool,
        target_name: &str,
    ) {
        self.infer(decorator);

        let dec_name = match decorator {
            Expr::Ident(name) => name.clone(),
            _ => return,
        };

        let expected_what = if target_is_fn { "function" } else { "type" };
        let target_kind = if target_is_fn { "function" } else { "class" };

        let is_fn_type = |ty: &InferredType| matches!(ty, InferredType::Function { .. });
        let is_type_type =
            |ty: &InferredType| matches!(ty, InferredType::TypeVal | InferredType::TypeValOf(_));
        let kind_matches = |ty: &InferredType| {
            if target_is_fn {
                is_fn_type(ty)
            } else {
                is_type_type(ty)
            }
        };

        // --- Case 1: 関数デコレータ ---
        if let Some(sigs) = self.fn_sigs.get(&dec_name).cloned() {
            if sigs.len() != 1 {
                return;
            }
            let sig = sigs[0].clone();

            match sig.params.first() {
                None => {
                    self.report_error(StaticTypeError {
                        kind: TypeErrorKind::InvalidDecorator {
                            reason: format!(
                                "decorator '{dec_name}' applied to {target_kind} '{target_name}': \
                                 must have at least one parameter of '{expected_what}' type"
                            ),
                        },
                        span: None,
                    });
                }
                Some((_, Some(first_param_ty))) => {
                    if !kind_matches(first_param_ty) {
                        self.report_error(StaticTypeError {
                            kind: TypeErrorKind::InvalidDecorator {
                                reason: format!(
                                    "decorator '{dec_name}' applied to {target_kind} '{target_name}': \
                                     first parameter must be '{expected_what}' type, got '{first_param_ty}'"
                                ),
                            },
                            span: None,
                        });
                    }
                }
                Some((_, None)) => {}
            }

            if let Some(return_ty) = &sig.return_type.clone() {
                if !kind_matches(return_ty) {
                    self.report_error(StaticTypeError {
                        kind: TypeErrorKind::InvalidDecorator {
                            reason: format!(
                                "decorator '{dec_name}' applied to {target_kind} '{target_name}': \
                                 return type must be '{expected_what}', got '{return_ty}'"
                            ),
                        },
                        span: None,
                    });
                }
            }
            return;
        }

        // --- Case 2: クラスデコレータ ---
        if self.known_class_names.contains(dec_name.as_str()) {
            let cls_methods = match self.class_method_sigs.get(&dec_name).cloned() {
                Some(m) => m,
                None => return,
            };

            if let Some(init_sigs) = cls_methods.get("__init__").cloned() {
                if init_sigs.len() == 1 {
                    if let Some((_, second_ty_opt)) = init_sigs[0].params.get(1) {
                        if let Some(second_ty) = second_ty_opt {
                            if !kind_matches(second_ty) {
                                self.report_error(StaticTypeError {
                                    kind: TypeErrorKind::InvalidDecorator {
                                        reason: format!(
                                            "class decorator '{dec_name}' applied to {target_kind} '{target_name}': \
                                             '__init__' second parameter must be '{expected_what}' type, got '{second_ty}'"
                                        ),
                                    },
                                    span: None,
                                });
                            }
                        }
                    }
                }
            }

            if let Some(call_sigs) = cls_methods.get("__call__").cloned() {
                if call_sigs.len() == 1 {
                    if let Some(return_ty) = &call_sigs[0].return_type.clone() {
                        if !kind_matches(return_ty) {
                            self.report_error(StaticTypeError {
                                kind: TypeErrorKind::InvalidDecorator {
                                    reason: format!(
                                        "class decorator '{dec_name}' applied to {target_kind} '{target_name}': \
                                         '__call__' return type must be '{expected_what}', got '{return_ty}'"
                                    ),
                                },
                                span: None,
                            });
                        }
                    }
                }
            }
        }
    }

    /// 型ガード式の右辺に書かれた型名文字列を [`InferredType`] に変換する。
    pub(super) fn type_from_guard_name(name: &str) -> InferredType {
        match name {
            "int" => InferredType::Int,
            "float" => InferredType::Float,
            "str" => InferredType::Str,
            "bool" => InferredType::Bool,
            "None" => InferredType::None,
            "Undefined" => InferredType::Undefined,
            other => InferredType::NamedInstance(other.to_string()),
        }
    }
}
